### PR TITLE
Support Scala.js and further reorg (part2)

### PR DESCRIPTION
**Pull request by aslesarenko**
In this PR:
- [x] sigmajs-crypto-facade JS dependency imported 
- [x] CryptoFacade JS implementation
- [x] Consensus code is further separated from compiler
- [x] More tests moved to `interpreter` module and run under both JVM and JS
- [x] some classes exposed to JS via JSExport annotation

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -235,11 +235,23 @@ lazy val interpreter = crossProject(JVMPlatform, JSPlatform)
...
+//      ("org.scala-js" %%% "scalajs-java-securerandom" % "1.0.0").cross(CrossVersion.for3Use2_13)
     ),
     useYarn := true
   )
 lazy val interpreterJS = interpreter.js
     .enablePlugins(ScalaJSBundlerPlugin)
+    .settings(
+      scalaJSLinkerConfig ~= { conf =>
+        conf.withSourceMap(false)
+//          .withModuleKind(ModuleKind.CommonJSModule)
```
Comment (kushti) [inline build.sbt:247]: commented out code

File: build.sbt [hunk 2]
```diff
@@ -235,11 +235,23 @@ lazy val interpreter = crossProject(JVMPlatform, JSPlatform)
...
+      },
+
+      //      scalaJSUseMainModuleInitializer := true,
```
Comment (kushti) [inline build.sbt:250]: commented out code

File: interpreter/js/src/main/scala/sigmastate/js/ErgoTree.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+import scorex.util.encode.Base16
+import sigmastate.Values
+import sigmastate.serialization.ErgoTreeSerializer
+
+import scala.scalajs.js
+import scala.scalajs.js.annotation.{JSExport, JSExportTopLevel}
+
+@JSExportTopLevel("ErgoTree")
+class ErgoTree(tree: Values.ErgoTree) extends js.Object {
```
Comment (kushti) [inline interpreter/js/src/main/scala/sigmastate/js/ErgoTree.scala:11]: ScalaDocs missed

File: interpreter/js/src/main/scala/sigmastate/js/Isos.scala [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package sigmastate.js
+
+import sigmastate.{Iso, SType}
+import sigmastate.Values.Constant
+import sigmastate.eval.Evaluation
+
+object Isos {
```
Comment (kushti) [inline interpreter/js/src/main/scala/sigmastate/js/Isos.scala:7]: ScalaDocs missed in the file

File: interpreter/js/src/main/scala/sigmastate/js/Interpreter.scala [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package sigmastate.js
+
+import scala.scalajs.js.annotation.JSExportTopLevel
+
+@JSExportTopLevel("Interpreter")
+object Interpreter {
```
Comment (kushti) [inline interpreter/js/src/main/scala/sigmastate/js/Interpreter.scala:6]: ScalaDoc missed

**Review by kushti — CHANGES_REQUESTED** [review]

File: common/shared/src/main/scala/sigmastate/Iso.scala [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  def to(a: A): B
+  def from(b: B): A
+  def andThen[C](iso: Iso[B,C]): Iso[A,C] = ComposeIso(iso, this)
+  def inverse: Iso[B, A] = InverseIso(this)
+}
+final case class InverseIso[A,B](iso: Iso[A,B]) extends Iso[B,A] {
+  override def to(a: B): A = iso.from(a)
+  override def from(b: A): B = iso.to(b)
+}
+final case class ComposeIso[A, B, C](iso2: Iso[B, C], iso1: Iso[A, B]) extends Iso[A, C] {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply common/shared/src/main/scala/sigmastate/Iso.scala:25]: fixed

File: common/shared/src/main/scala/sigmastate/Iso.scala [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  def from(c: C): A = iso1.from(iso2.from(c))
+  def to(a: A): C = iso2.to(iso1.to(a))
+}
+
+trait LowPriorityIsos {
+}
+
+object Iso extends LowPriorityIsos {
+  implicit def identityIso[A]: Iso[A, A] = new Iso[A, A] {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply common/shared/src/main/scala/sigmastate/Iso.scala:34]: fixed

File: common/js/src/main/scala/scalan/reflection/Platform.scala [hunk 8]
```diff
@@ -22,4 +24,27 @@ object Platform {
     }
     res.asInstanceOf[RClass[T]]
   }
+
+  /** A cache that stores key-value pairs using a synchronized HashMap.
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply common/js/src/main/scala/scalan/reflection/Platform.scala:28]: fixed

File: common/js/src/main/scala/scalan/reflection/Platform.scala [hunk 9]
```diff
@@ -22,4 +24,27 @@ object Platform {
...
+    *
+    * @tparam K the type of keys used in the cache
+    * @tparam V the type of values stored in the cache
+    */
+  class Cache[K, V] {
+
+    /** key-value pairs. */
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply common/js/src/main/scala/scalan/reflection/Platform.scala:35]: fixed

File: interpreter/js/src/main/scala/sigmastate/crypto/Imported.scala [hunk 10]
```diff
@@ -0,0 +1,60 @@
+package sigmastate.crypto
+
+import scala.scalajs.js
+import scala.scalajs.js.annotation.JSImport
+import scala.scalajs.js.typedarray.Uint8Array
+
+@js.native
+@JSImport("sigmajs-crypto-facade", "CryptoContext")
+class CryptoContextJs() extends js.Object {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply interpreter/js/src/main/scala/sigmastate/crypto/Imported.scala:9]: fixed

File: build.sbt (hunk 1 above)
Comment (aslesarenko) [reply build.sbt:247]: fixed

File: build.sbt (hunk 2 above)
Comment (aslesarenko) [reply build.sbt:250]: fixed

File: interpreter/js/src/main/scala/sigmastate/js/ErgoTree.scala (hunk 3 above)
Comment (aslesarenko) [reply interpreter/js/src/main/scala/sigmastate/js/ErgoTree.scala:11]: fixed

File: interpreter/js/src/main/scala/sigmastate/js/Isos.scala (hunk 4 above)
Comment (aslesarenko) [reply interpreter/js/src/main/scala/sigmastate/js/Isos.scala:8]: fixed

File: interpreter/js/src/main/scala/sigmastate/js/Interpreter.scala (hunk 5 above)
Comment (aslesarenko) [reply interpreter/js/src/main/scala/sigmastate/js/Interpreter.scala]: removed in this PR, will be added later.

**Review by kushti — APPROVED** [review]

**Pull request merged**